### DISCUSSION TITLE
How to repeat exact Photoshop actions in Magick .net to apply a mask with proper feathering on edges?

### DISCUSSION THREAD

**Discussion by futureimage** [ImageMagick/ImageMagick#7031] [Digital Image Processing]
I have two images: one color image (original) and one grayscale image created by AI with only black and white colors to work as mask. I have combined these as one layered image:

![Screenshot 2024-01-15 at 12 27 09](https://github.com/ImageMagick/ImageMagick/assets/7640227/d65e8ed9-df12-4a95-800c-474060f7de7d)

I am applying the white area as mask to make the background transparent. In photoshop:

Select -> Color range (White), Fuzziness 100
Select -> Modify -> feather 1px
Layer -> Layer Mask -> Reveal Selection

![Screenshot 2024-01-15 at 12 31 14](https://github.com/ImageMagick/ImageMagick/assets/7640227/b48c3201-1da6-4e01-95a1-e2d7e9064b21)

The image transparency looks good with soft edges (red bg added for better visual evaluation).

![Screenshot 2024-01-15 at 12 22 56](https://github.com/ImageMagick/ImageMagick/assets/7640227/d9587c0f-01a8-4765-aaad-f7bbddce193b)

I am able to create the masked PNG file with Magick .Net, but the edges are not as good as they are with photoshop. This is my latest code:

```
using (var jpg = new MagickImage(original))
{
      using var png = new MagickImage(mask);
       png.Transparent(MagickColors.White);
       png.Alpha(AlphaOption.Set);
       png.Threshold(new Percentage(28));
       png.Morphology(MorphologyMethod.Erode, Kernel.Diamond, 1);
       png.Blur(1,0);

      jpg.Composite(png, CompositeOperator.CopyAlpha);
}
```

How can I repeat the Photoshop steps EXACTLY the same in Magick .Net?

Input images as zip: 
[files.zip](https://github.com/ImageMagick/ImageMagick/files/13946846/files.zip)

**Comment by fmw42** [comment]
Post your input images

**Reply by futureimage** [reply]
Due to copyright I can not post the image seen in my original question. I posted another image instead to the original question. Photoshop version on left, Magick.NET on right. The difference is not big, but it is visible.

![Screenshot 2024-01-16 at 9 48 30](https://github.com/ImageMagick/ImageMagick/assets/7640227/23693c68-2ff2-4697-97e1-0d49ac8b0f44)

**Comment by fmw42** [comment]
We need your separate input image, mask image and corresponding separate output from Photoshop.

**Reply by futureimage** [reply]
Input image sent to AI: 
![cropped](https://github.com/ImageMagick/ImageMagick/assets/7640227/8ed429cb-072d-4c3b-868b-3e17c85929c8)

Returned files: 
[cropped.zip](https://github.com/ImageMagick/ImageMagick/files/13955159/cropped.zip)

Magick.NET applied mask: 
![cropped](https://github.com/ImageMagick/ImageMagick/assets/7640227/65bbcac0-5ccc-44d3-8008-8869308d1e57)

```
png.ColorFuzz = new Percentage(30);
png.Transparent(MagickColors.Black);
png.Blur(0,1);
masked.Format = MagickFormat.Png;
masked.Composite(png, CompositeOperator.CopyAlpha);
```
Photoshop image can not be sent (21MB) but you can reproduce it using the workflow from the original question (layer 0 = color.jpg, layer 1 = alpha.png from the AI returned zip file):

Select -> Color range (White), Fuzziness 100 -> layer 1
Select -> Modify -> feather 1px -> selection
Layer -> Layer Mask -> Reveal Selection -> layer 0

**Comment by fmw42** [comment]
I have Photoshop, but am not an expert and do not know how you achieved the steps you show.  Please list step by step the menu/menu item that you used to achieve your results.  I do not know how you got to your image 

![296728303-b48c3201-1da6-4e01-95a1-e2d7e9064b21](https://github.com/ImageMagick/ImageMagick/assets/13458218/b8c3e9e1-ef38-42d8-bc8c-2480de48798c)


I suspect the issue is how Photoshop does it feathering. It may be different from how Imagemagick can do that.

In Imagemagick command line, you can do a basic feather and the rest of your processing as

Unix syntax:

```
magick color.jpg \
\( alpha.png -blur 0x1 -level 50x100% \) \
-alpha off -compose copy_opacity -composite \
-background red -alpha background -alpha off \
result3.png

```

![result3](https://github.com/ImageMagick/ImageMagick/assets/13458218/832bc4ef-84b5-40ac-9864-3e35f4d1e60b)

**Reply by futureimage** [reply]
Many thanks for your effort! The issue here is exactly how to produce the feather between opaque and transparent areas in IM like Photoshop does it. So being able to compare the IM result with the PS result is vital in solving it :)

I try to open the Photoshop workflow better below.

Show layers palette by selecting Windows -> Layers from the menu
Open both files from zip (color.jpg, alpha.png)
First move the alpha.png to color.jpg as own layer (hold shift (to place it exactly in same position than bg layer) and drag and drop the layer from png to jpg)

Then from the menus:
Select -> Color range (White), Fuzziness 100 (click anywhere on the white area on layer 1 with the pipette to select the white) -> Ok
Select -> Modify -> feather radius 1px -> Ok
Hide layer 1 and activate (select) layer 0 in Layers palette
Layer -> Layer Mask -> Reveal Selection

For the coloured background just add a new layer, drag it to bottom on layers palette and fill with wanted colour.

**Comment by snibgo** [comment]
@futureimage: your "feather radius" is just 1 pixel. With such a small amount, it is difficult to see exactly what is happening, and difficult to replicate the effect.

You might try a larger effect, eg a radius of 10 pixels.

**Comment by fmw42** [comment]
This is what I get in Photoshop when using your recipe above.

![color_ps_fred](https://github.com/ImageMagick/ImageMagick/assets/13458218/eb4c388a-b3eb-4b6b-852f-e3f5487daab3)


And this is what I get in Imagemagick with the following command

```
convert color.jpg \
\( alpha.png -blur 0x1 -level 50x100% \) \
-alpha off -compose copy_opacity -composite \
-compose over -background red -flatten \
result3b.png
```

![result3b](https://github.com/ImageMagick/ImageMagick/assets/13458218/3a7a9e4d-2974-402e-a83b-32757e6245bc)


Is this close enough?

**Reply by futureimage** [reply]
Your IM version looks very good! I am using Magick.NET and am not able to translate this and get the result you have. This is my code (masked = color.jpg, png = alpha.png):

```
using (var masked = new MagickImage(original))
{
        using var png = new MagickImage(mask);
        png.Blur(0, 1);
        png.Level(50, 100);
        png.Alpha(AlphaOption.Off);
        masked.Format = MagickFormat.Png;
        masked.Composite(png, CompositeOperator.CopyAlpha);
        await masked.WriteAsync(final);
}
```
This is my result:
![cropped copy](https://github.com/ImageMagick/ImageMagick/assets/7640227/6a9f0e34-c136-46ae-9016-567944816f07)

And btw I save the file as transparent png and just add later the red bg to visually see the produced transparency.

**Comment by snibgo** [comment]
The solution given by fmw42 introduces a darkening at the feathering. This is particularly obvious at top-right. Windows BAT script:
```
magick color.jpg ^
( alpha.png -blur 0x1 -level 50x100%% ) ^
-alpha off -compose copy_opacity -composite ^
-compose over -background red -flatten ^
-crop 131x150+2995+0 +repage ^
-scale 300%% ^
result3b.png
```
![result3b](https://github.com/ImageMagick/ImageMagick/assets/33812368/8795b397-aad7-464d-be61-7902f909e3a3)

The darkening is because the colours (the light cloth and the red bacground) are mixed at the feathering in sRGB space. A remedy is to convert both colours to RGB, then mix, then convert the result back to sRGB.

"red" has the same pixel values in sRGB and RGB, so we don't need to convert that.
```
magick color.jpg ^
-colorspace RGB ^
( alpha.png -blur 0x1 -level 50x100%% ) ^
-alpha off -compose copy_opacity -composite ^
-compose over -background red -flatten ^
-colorspace sRGB ^
-crop 131x150+2995+0 +repage ^
-scale 300%% ^
result3c.png
```
![result3c](https://github.com/ImageMagick/ImageMagick/assets/33812368/f9b13bb6-6f3a-449b-8fda-8c9ab260f8c6)

**Reply by futureimage** [reply]
Good observations! Perhaps this would work on all colors and backgrounds? If I only could write the same for Magick.NET :)
